File: apps/resnet_py/im_cls.py
import os
import sys
import argparse
sys.path.append('../')
import gi
gi.require_version('Gst', '1.0')
from gi.repository import GLib, Gst
from common.platform_info import PlatformInfo
from common.bus_call import bus_call

import pyds
from resnet_parser import nvds_infer_parse_custom_resnet

MUXER_BATCH_TIMEOUT_USEC = 33000

def pgie_src_pad_buffer_probe(pad, info, u_data):
    gst_buffer = info.get_buffer()
    if not gst_buffer:
        print("Unable to get GstBuffer ")
        return Gst.PadProbeReturn.OK

    # Retrieve batch metadata from the GstBuffer
    batch_meta = pyds.gst_buffer_get_nvds_batch_meta(hash(gst_buffer))

    l_frame = batch_meta.frame_meta_list
    while l_frame is not None:
        try:
            frame_meta = pyds.NvDsFrameMeta.cast(l_frame.data)
        except StopIteration:
            break

        l_user_meta = frame_meta.frame_user_meta_list
        frame_number = frame_meta.frame_num
        output_class = "class_fake"
        while l_user_meta is not None:
            try:
                user_meta = pyds.NvDsUserMeta.cast(l_user_meta.data)
            except StopIteration:
                break
            if user_meta.base_meta.meta_type == pyds.NvDsMetaType.NVDSINFER_TENSOR_OUTPUT_META:
                # Cast user_meta_data to NvDsInferTensorMeta
                tensor_meta = pyds.NvDsInferTensorMeta.cast(user_meta.user_meta_data)
                output_class = nvds_infer_parse_custom_resnet(tensor_meta)
            try:
                l_user_meta = l_user_meta.next
            except StopIteration:
                break
        # Get or create NvDsDisplayMeta for the current frame
        display_meta = pyds.nvds_acquire_display_meta_from_pool(batch_meta)
        if display_meta is None:
            print("Failed to acquire NvDsDisplayMeta from pool. Skipping display.")
            try:
                l_frame = l_frame.next
            except StopIteration:
                break
            continue

        # Create NvOSD_TextParams for your string
        display_meta=pyds.nvds_acquire_display_meta_from_pool(batch_meta) #Retrieve NvDsDisplayMeta object from given NvDsBatchMeta object. See NvDsMeta docs for more details.
        display_meta.num_labels = 1
        py_nvosd_text_params = display_meta.text_params[0] #Retrieve NvOSD_TextParams object from list in display meta. 
        # Setting display text to be shown on screen
        # Note that the pyds module allocates a buffer for the string, and the
        # memory will not be claimed by the garbage collector.
        # Reading the display_text field here will return the C address of the
        # allocated string. Use pyds.get_string() to get the string content.
        py_nvosd_text_params.display_text = f"Frame Number={frame_number} Class={output_class}"

        # Now set the offsets where the string should appear
        py_nvosd_text_params.x_offset = 10
        py_nvosd_text_params.y_offset = 12

        # Font , font-color and font-size
        py_nvosd_text_params.font_params.font_name = "Serif" #Set attributes of our NvOSD_TextParams object's NvOSD_FontParams member 
        py_nvosd_text_params.font_params.font_size = 10
        # set(red, green, blue, alpha); set to White
        py_nvosd_text_params.font_params.font_color.set(1.0, 1.0, 1.0, 1.0) #See NvOSD_ColorParams for more details.

        # Text background color
        py_nvosd_text_params.set_bg_clr = 1 #Set boolean indicating that text has bg color to true. 
        # set(red, green, blue, alpha); set to Black
        py_nvosd_text_params.text_bg_clr.set(0.0, 0.0, 0.0, 1.0)
        # Using pyds.get_string() to get display_text as string
        print(pyds.get_string(py_nvosd_text_params.display_text))
        pyds.nvds_add_display_meta_to_frame(frame_meta, display_meta) #Add display meta to frame after setting text params attributes.

        try:
            l_frame = l_frame.next
        except StopIteration:
            break

    return Gst.PadProbeReturn.OK

def osd_sink_pad_buffer_probe(pad,info,u_data):
    gst_buffer = info.get_buffer()
    if not gst_buffer:
        print("Unable to get GstBuffer ")
        return

    # Retrieve batch metadata from the gst_buffer
    # Note that pyds.gst_buffer_get_nvds_batch_meta() expects the
    # C address of gst_buffer as input, which is obtained with hash(gst_buffer)
    batch_meta = pyds.gst_buffer_get_nvds_batch_meta(hash(gst_buffer))
    l_frame = batch_meta.frame_meta_list
    while l_frame is not None:
        try:
            # Note that l_frame.data needs a cast to pyds.NvDsFrameMeta
            # The casting is done by pyds.NvDsFrameMeta.cast()
            # The casting also keeps ownership of the underlying memory
            # in the C code, so the Python garbage collector will leave
            # it alone.
            frame_meta = pyds.NvDsFrameMeta.cast(l_frame.data)
        except StopIteration:
            break

        try:
            l_frame=l_frame.next
        except StopIteration:
            break
			
    return Gst.PadProbeReturn.OK	


def main(args):
    pgie_type = args.pgie_type
    input_video = args.input_video
    
    platform_info = PlatformInfo()
    # Init Gstreamer
    Gst.init(None)

    # Create gstreamer pipeline
    print("Creating pipeline")
    pipeline = Gst.Pipeline()
    if not pipeline:
        sys.stderr.write("Can not create pipeline")

    # Create element to read from file
    print("Create source")
    source = Gst.ElementFactory.make("filesrc", "file-source")
    if not source:
        sys.stderr.write("Unable to create file source")

    # Create elements to form pipeline in gstreamer
    # Input file is a single h264 encoded file
    # So we needs to h264parser
    print("Create H264 Parser")
    h264parser = Gst.ElementFactory.make("h264parse", "h264-parser")
    if not h264parser:
        sys.stderr.write("Unable to create h264 parser")
    
    # use nvdec_h264 for hardware accelerated decode on GPU
    print("Create decoder")
    decoder = Gst.ElementFactory.make("nvv4l2decoder", "nvv4l2-decoder")
    if not decoder:
        sys.stderr.write("Unable to create Nv v4l2 decoder")
    
    # Create nvstreammux instance to form batches from one or more sources.
    streammux = Gst.ElementFactory.make("nvstreammux", "Stream-muxer")
    if not streammux:
        sys.stderr.write(" Unable to create NvStreamMux \n")

    if pgie_type == "nvinfer":
        # Use nvinfer to run inferencing on decoder's output,
        # behaviour of inferencing is set through config file
        pgie = Gst.ElementFactory.make("nvinfer", "primary-inference")
        if not pgie:
            sys.stderr.write(" Unable to create pgie with nvinfer \n")
    else:
        pgie = Gst.ElementFactory.make("nvinferserver", "primary-inference")
        if not pgie:
            sys.stderr.write(" Unable to create pgie with nvinfer-server \n")
    # Use convertor to convert from NV12 to RGBA as required by nvosd
    nvvidconv = Gst.ElementFactory.make("nvvideoconvert", "convertor")
    if not nvvidconv:
        sys.stderr.write(" Unable to create nvvidconv \n")

    # Create OSD to draw on the converted RGBA buffer
    nvosd = Gst.ElementFactory.make("nvdsosd", "onscreendisplay")
    if not nvosd:
        sys.stderr.write(" Unable to create nvosd \n")

    # Finally render the osd output
    if platform_info.is_integrated_gpu():
        print("Creating nv3dsink \n")
        sink = Gst.ElementFactory.make("nv3dsink", "nv3d-sink")
        if not sink:
            sys.stderr.write(" Unable to create nv3dsink \n")
    else:
        if platform_info.is_platform_aarch64():
            print("Creating nv3dsink \n")
            sink = Gst.ElementFactory.make("nv3dsink", "nv3d-sink")
        else:
            print("Creating EGLSink \n")
            sink = Gst.ElementFactory.make("nveglglessink", "nvvideo-renderer")
        if not sink:
            sys.stderr.write(" Unable to create egl sink \n")

    print(f"Playing file {input_video}")
    source.set_property('location', input_video)
    if os.environ.get('USE_NEW_NVSTREAMMUX') != 'yes': # Only set these properties if not using new gst-nvstreammux
        streammux.set_property('width', 1920)
        streammux.set_property('height', 1080)
        streammux.set_property('batched-push-timeout', MUXER_BATCH_TIMEOUT_USEC)

    streammux.set_property('batch-size', 1)
    if pgie_type == "nvinfer":
        pgie.set_property('config-file-path', "nvinfer_config.txt")
    else:
        pgie.set_property('config-file-path', "nvinfer_server_config.txt")

    print("Adding element to Pipeline ")
    pipeline.add(source)
    pipeline.add(h264parser)
    pipeline.add(decoder)
    pipeline.add(streammux)
    pipeline.add(pgie)
    pipeline.add(nvvidconv)
    pipeline.add(nvosd)
    pipeline.add(sink)

    # we link the elements together
    # file-source -> h264-parser -> nvh264-decoder ->
    # nvinfer -> nvvidconv -> nvosd -> video-renderer
    print("Linking elements in the Pipeline \n")
    source.link(h264parser)
    h264parser.link(decoder)

    sinkpad = streammux.request_pad_simple("sink_0")
    if not sinkpad:
        sys.stderr.write(" Unable to get the sink pad of streammux \n")
    srcpad = decoder.get_static_pad("src")
    if not srcpad:
        sys.stderr.write(" Unable to get source pad of decoder \n")
    srcpad.link(sinkpad)
    streammux.link(pgie)
    pgie.link(nvvidconv)
    nvvidconv.link(nvosd)
    nvosd.link(sink)

    # create an event loop and feed gstreamer bus mesages to it
    loop = GLib.MainLoop()
    bus = pipeline.get_bus()
    bus.add_signal_watch()
    bus.connect ("message", bus_call, loop)

    # Add a probe on the primary-infer soruce pad to get infer output tensors
    pgie_src_pad = pgie.get_static_pad("src")
    if not pgie_src_pad:
        sys.stderr.write(" Unable to get src pad of primary infer engine")
    else:
        pgie_src_pad.add_probe(Gst.PadProbeType.BUFFER, pgie_src_pad_buffer_probe, 0)

    # Lets add probe to get informed of the meta data generated, we add probe to
    # the sink pad of the osd element, since by that time, the buffer would have
    # had got all the metadata.
    osdsinkpad = nvosd.get_static_pad("sink")
    if not osdsinkpad:
        sys.stderr.write(" Unable to get sink pad of nvosd \n")

    osdsinkpad.add_probe(Gst.PadProbeType.BUFFER, osd_sink_pad_buffer_probe, 0)

    # start play back and listen to events
    print("Starting pipeline \n")
    pipeline.set_state(Gst.State.PLAYING)
    try:
        loop.run()
    except:
        pass
    # cleanup
    pipeline.set_state(Gst.State.NULL)

def get_args():
    parser = argparse.ArgumentParser(prog="sample resnet")
    parser.add_argument("--pgie_type", type=str, choices=["nvinfer", "nvinfer-server"], default="nvinfer")
    parser.add_argument("--input_video", type=str, default="/opt/nvidia/deepstream/deepstream/samples/streams/sample_720p.h264")
    return parser.parse_args()

if __name__ == '__main__':
    args = get_args()
    sys.exit(main(args))

File: apps/resnet_cpp/im_cls.cpp
#include <gst/gst.h>
#include <glib.h>

#include <memory>
#include <cassert>

#include <cuda_runtime_api.h>
#include "gstnvdsmeta.h"

static GstPadProbeReturn osd_sink_pad_buffer_probe(GstPad *pad, GstPadProbeInfo *info, gpointer u_data) {
    GstBuffer *buf=(GstBuffer*)info->data;
    NvDsMetaList * l_frame = NULL;
    NvDsMetaList * l_obj = NULL;
    NvDsDisplayMeta *display_meta = NULL;
}

static gboolean bus_call(GstBus *bus, GstMessage *msg, gpointer data) {
    GMainLoop *loop = (GMainLoop*) data;
    switch (GST_MESSAGE_TYPE(msg))
    {
    case GST_MESSAGE_EOS:
        g_print("End of stream \n");
        g_main_loop_quit(loop);
        break;
    case GST_MESSAGE_ERROR:
        {
            gchar* debug;
            GError *error;
            gst_message_parse_error(msg, &error, &debug);
            g_printerr("Error from element %s: %s \n", 
                GST_OBJECT_NAME(msg->src), error->message
            );
            if (debug)
                g_printerr("Error details: %s \n", debug);
            g_free(debug);
            g_error_free(error);
            g_main_loop_quit(loop);
            break;
        }
    default:
        break;
    }
    return true;
}


int main(int argc, char *argv[]) {
    GMainLoop *loop=NULL;
    GstElement *pipeline=NULL, *source=NULL, *h264parser=NULL, *decoder=NULL, 
    *streammux=NULL, *pgie=NULL, *nvvidconv=NULL, *nvosd=NULL, *sink=NULL;

    int current_device=-1;
    cudaGetDevice(&current_device);
    struct cudaDeviceProp prop;
    cudaGetDeviceProperties(&prop, current_device);
    const gchar* new_mux_str = g_getenv("USE_NEW_NVSTREAMMUX");
    gboolean use_new_mux = !g_strcmp0(new_mux_str, "yes");
    
    GstBus *bus;
    guint bus_watch_id = 0;
    GstPad *osd_sink_pad=NULL;

    // Stanard Gstreamer initialization
    gst_init(&argc, &argv);
    loop = g_main_loop_new(NULL, false);

    // Create gstreamers elements
    // Create Pipeline elemetsn to connect other elements
    pipeline = gst_pipeline_new("im-cls-pipeline");
    if(!pipeline) {
        g_printerr("Can't create pipeline. Exiting \n");
        return -1;
    }

    // Create element filesrc to read from file 
    source = gst_element_factory_make("filesrc", "file-source");
    if(!source) {
        g_printerr("Can't create filesrc to read h264 file. Exiting \n");
        return -1;
    }

    // Input file is a single h264 encoded file
    // So we needs to h264parser
    // Create element h264parse to parse h264 stream
    h264parser = gst_element_factory_make("h264parse", "h264-parser");
    if(!source) {
        g_printerr("Can't create h264parse to parse h264 file. Exiting \n");
        return -1;
    }

    // use nvdec_h264 for hardware accelerated decode on GPU
    decoder = gst_element_factory_make("nvv4l2decoder", "nvv4l2-decoder");
    if(!decoder) {
        g_printerr("Can't create Nv v4l2 decoder. Exiting \n");
        return -1;
    }

    // Create nvstreammux instance to form batches from one or more sources.
    streammux = gst_element_factory_make("nvstreammux", "Stream-muxer");
    if(!streammux) {
        g_printerr("Unable to create NvStreamMux. Exiting \n");
        return -1;
    }

    // Use nvinfer to run inferencing on decoder's output,
    // behaviour of inferencing is set through config file
    pgie = gst_element_factory_make("nvinfer", "primary-nvinference-engine");
    if(!pgie) {
        g_printerr("Unable to create pgie with nvinfer. Exiting \n");
        return -1;
    }

    // Use convertor to convert from NV12 to RGBA as required by nvosd
    nvvidconv = gst_element_factory_make("nvvideoconvert", "convertor");
    if(!nvvidconv) {
        g_printerr("Unable to create nvvidconvr. Exiting \n");
        return -1;
    }

    // Create OSD to draw on the converted RGBA buffer
    nvosd = gst_element_factory_make("nvdsosd", "onscreendisplay");
    if(!nvosd) {
        g_printerr("Unable to create nvosd. Exiting \n");
        return -1;
    }

    // Finally render the osd output
    if(prop.integrated) {
        sink = gst_element_factory_make("nv3dsink", "nv3d-sink");
        if(!sink) {
            g_printerr("Unable to create nv3dsink. Exiting \n");
            return -1;
        }
    } 
    else {
    #ifdef __aarch64__
        sink = gst_element_factory_make ("nv3dsink", "nvvideo-renderer");
    #else
        sink = gst_element_factory_make ("nveglglessink", "nvvideo-renderer");
        if(!sink) {
            g_printerr("Unable to create eglsink. Exiting \n");
            return -1;
        }
    #endif
    }
    
    // Settting properties of elements here
    // set h264 file to run
    g_object_set(G_OBJECT(source), "location", argv[1], NULL);

    g_object_set(G_OBJECT(streammux), "batch-size", 1, NULL);
    g_object_set(G_OBJECT(streammux), "width", 1920, NULL);
    g_object_set(G_OBJECT(streammux), "height", 1080, NULL);
    g_object_set(G_OBJECT(streammux), "batched-push-timeout", 100000, NULL);

    g_object_set(G_OBJECT(pgie), "config-file-path", "nvinfer_config.txt", NULL);

    // Message handler
    bus = gst_pipeline_get_bus (GST_PIPELINE(pipeline));
    bus_watch_id = gst_bus_add_watch(bus, bus_call, loop);
    gst_object_unref(bus);

    // setup pipeline
    // We can connect elements one by one as in python version
    // but c version allow us to connect multiple element at once too
    gst_bin_add_many(GST_BIN(pipeline),
      source, h264parser, decoder, streammux, pgie,
      nvvidconv, nvosd, sink, NULL);

    // Pad to link decoder and streammux
    GstPad *sinkpad, *srcpad;
    gchar pad_name_sink[16] = "sink_0";
    gchar pad_name_src[16] = "src";

    sinkpad = gst_element_request_pad_simple (streammux, pad_name_sink);
    if (!sinkpad) {
        g_printerr ("Streammux request sink pad failed. Exiting.\n");
        return -1;
    }

    srcpad = gst_element_get_static_pad (decoder, pad_name_src);
    if (!srcpad) {
        g_printerr ("Decoder request src pad failed. Exiting.\n");
        return -1;
    }

    if (gst_pad_link (srcpad, sinkpad) != GST_PAD_LINK_OK) {
        g_printerr ("Failed to link decoder to stream muxer. Exiting.\n");
        return -1;
    }

    gst_object_unref (sinkpad);
    gst_object_unref (srcpad);

    // Link elements together, decoder and streammux are linked by pad in previous step
    if (!gst_element_link_many (source, h264parser, decoder, NULL)) {
        g_printerr ("Elements could not be linked: 1. Exiting.\n");
        return -1;
    }
    
    if (!gst_element_link_many (streammux, pgie, nvvidconv, nvosd, sink, NULL)) {
        g_printerr ("Elements could not be linked: 2. Exiting.\n");
        return -1;
    }

    // create pad to parse output of ods 
    osd_sink_pad = gst_element_get_static_pad (nvosd, "sink");
    if (!osd_sink_pad)
        g_print ("Unable to get sink pad\n");
    else
        gst_pad_add_probe (osd_sink_pad, GST_PAD_PROBE_TYPE_BUFFER,
            osd_sink_pad_buffer_probe, NULL, NULL);
    gst_object_unref (osd_sink_pad);

    /* Set the pipeline to "playing" state */
    g_print ("Using file: %s\n", argv[1]);
    gst_element_set_state (pipeline, GST_STATE_PLAYING);

    /* Wait till pipeline encounters an error or EOS */
    g_print ("Running...\n");
    g_main_loop_run (loop);

    /* Out of the main loop, clean up nicely */
    g_print ("Returned, stopping playback\n");
    gst_element_set_state (pipeline, GST_STATE_NULL);
    g_print ("Deleting pipeline\n");
    gst_object_unref (GST_OBJECT (pipeline));
    g_source_remove (bus_watch_id);
    g_main_loop_unref (loop);

    return 0;
}